src/python/zquantum/core/cost_function.py

```diff
@@ -275,3 +277,72 @@ def __call__(self, parameters: np.ndarray) -> ValueEstimate:
         )
 
         return sum_expectation_values(combined_expectation_values)
+
+
+class CostFunction(Protocol):
+    """Cost function transforming vectors from R^n to numbers or their estimates."""
+
+    @abc.abstractmethod
+    def __call__(self, params: np.ndarray) -> Union[float, ValueEstimate]:
+        """Compute  value of the cost function for given parameters."""
+
+
+class EstimationTasksFactory(Protocol):
+    """Factory from producing estimation tasks from R^n vectors.
+
+    For instance, this can be used with ansatzes where produced estimation tasks
+    are evaluating circuit.
+    """
+
+    @abc.abstractmethod
+    def __call__(self, parameters: np.ndarray) -> List[EstimationTask]:
+        """Produce estimation tasks for given parameters."""
+
+
+class ParameterPreprocessor(Protocol):
+    """Parameter preprocessor.
+
+    Implementer of this protocol should create new array instead of
+    modifying passed parameters in place, which can have unpredictable
+    side effects.
+    """
+
+    @abc.abstractmethod
+    def __call__(self, parameters: np.ndarray) -> np.ndarray:
+        """Preprocess parameters."""
+
+
+def fix_parameters(fixed_parameters: np.ndarray) -> ParameterPreprocessor:
```

Comment:
Shouldn't it actually prepend instead of append?
For LBL training for example we:
- train a layer –> you have some good parameters for layer 1
- add a new layer
- use new parameter vector – good params from layer 1 + some new random params.